src/Qir/Runtime/lib/QIR/QubitManagerRestrictedReuse.hpp

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include <csignal>
+#include <iostream>
+#include <vector>
+
+#include "CoreTypes.hpp"
+
+namespace Microsoft
+{
+namespace Quantum
+{
+
+    // TODO: Remove this! Use standard ones.
+    static void FailNow(const char* message) { std::cout << message; throw message; }
+    static void FailIf(bool condition, const char* message) {
+        if (condition) {
+            FailNow(message);
+        }
+    }
+
+
+    // TODO: Decide on the interface and move this to appropriate place.
+    struct IQubitManager
+    {
+        virtual ~IQubitManager() = default;
+        virtual Qubit Allocate() = 0;
+        virtual void Release(Qubit qubit) = 0;
+    };
+
+    // The end of free lists are marked with NoneMarker value. It is used like null for pointers.
+    // This value is non-negative just like other values in the free lists.
+    constexpr int NoneMarker = INT_MAX;
+
+    // Explicitly allocated qubits are marked with Allocated value.
+    // Qubits implicitly allocated for borrowing are "refcounted" with values [Allocated+1 .. -2].
+    // When refcount needs to be decreased to Allocated value, the qubit is automatically released.
+    constexpr int AllocatedMarker = INT_MAX;
+
+    // Disabled qubits are marked with this value.
+    constexpr int DisabledMarker = -1;
+
+    // We want status array to be reasonably large if initialization for it is wrong (i.e. requested size < 1).
+    constexpr int FallbackQubitCapacity = 8;
+
+    // Indexes in the status array can potentially be in range 0 .. long.MaxValue-1.
+    // This gives maximum capacity as long.MaxValue.
+    // Index equal to long.MaxValue doesn't exist and is reserved for 'NoneMarker' - list terminator.
+    constexpr int MaximumQubitCapacity = INT_MAX;
+
+    // QubitListInSharedArray implements a singly-linked list with "pointers" to the first and the last element stored.
+    // Pointers are the indexes in a shared array. Shared array isn't sotored here because it can be reallocated.
+    // This class maintains status of elements in the list by virtue of linking them as part of this list.
+    // This class doesn't update status of elementes excluded from the list.
+    // This class is small, contains no pointers and relies on default shallow copying/destruction.
+    struct QubitListInSharedArray final
+    {
+    private:
+        int firstElement = NoneMarker;
+        int lastElement = NoneMarker;
+
+    public:
+        // Initialize empty list
+        QubitListInSharedArray() = default;
+
+        // Initialize as a list with sequential elements from minElement to maxElement inclusve.
+        QubitListInSharedArray(int minElement, int maxElement, int* sharedQubitStatusArray);
```

Comment:
(For the future)
From the description it is not evident how large should the array be, pointed to by the `sharedQubitStatusArray` (how many elements it should have - e.g. `(maxElement - minElement)` elements or other?). It is also not evident (from the description) that the elements `sharedQubitStatusArray[minElement]` through the `sharedQubitStatusArray[maxElement]` are _accessed_ in this function.

Also for me personally it is unexpected that the constructor updates anything _outside_ of the instance it constructs.

Reply:
I renamed parameters. This class may be moved to be a nested class of QubitManager. That way it would be easier to understand what's going on. However, as this is called "list in shared array" and this shared array is passed in, I think it's reasonable to assume that this will adjust the array.

Reply:
I will actually unresolve this as I'm still contemplating making auxiliary classes nested in CQubitManager.

Reply:
I moved all auxiliary classes, types and constants to be nested in CQubitManager. This made it more clear that these classes are bound to the Qubit Manager.